src/OneDriveSdk/Helpers/UploadSessionHelper.cs

```diff
@@ -0,0 +1,257 @@
+        public async Task<Item> UploadAsync(int maxTries = 3, IEnumerable<Option> options = null)
+        {
+            var uploadTries = 0;
+            var readBuffer = new byte[this.maxChunkSize];
+            
+            while (uploadTries < maxTries)
+            {
+                var chunkRequests = this.GetUploadChunkRequests(options);
+
+                foreach (var request in chunkRequests)
+                {
+                    var tries = 0;
+                    var requestSucceeded = false;
+                    this.uploadStream.Seek(request.RangeBegin, SeekOrigin.Begin);
+                    await this.uploadStream.ReadAsync(readBuffer, 0, request.RangeLength).ConfigureAwait(false);
+
+                    while (tries < 2 && !requestSucceeded) // Retry a given request only once
+                    {
+                        using (var requestBodyStream = new MemoryStream(request.RangeLength))
+                        {
+                            await requestBodyStream.WriteAsync(readBuffer, 0, request.RangeLength).ConfigureAwait(false);
+                            requestBodyStream.Seek(0, SeekOrigin.Begin);
+
+                            try
+                            {
+                                var result = await request.PutAsync(requestBodyStream).ConfigureAwait(false);
+                                if (result.UploadSucceeded)
+                                {
+                                    return result.ItemResponse;
+                                }
+
+                                requestSucceeded = true;
+                            }
+                            catch (ServiceException exception)
+                            {
+                                if (exception.IsMatch("generalException") || exception.IsMatch("timeout"))
+                                {
+                                    // Swallow and retry
+                                    tries += 1;
+                                }
+                                else if (exception.IsMatch("invalidRange"))
+                                {
+                                    // Range already received, so a previous request was successful
+                                    requestSucceeded = true;
+                                }
+                                else
+                                {
+                                    throw;
+                                }
+                            }
+                        }
+                    }
+                }
+
+                await this.UpdateSessionStatusAsync();
+                uploadTries += 1;
+                if (uploadTries < maxTries)
+                {
+                    // Exponential backoff in case of failures.
+                    await Task.Delay(2000 * uploadTries * uploadTries).ConfigureAwait(false);
+                }
+            }
+
+            throw new TaskCanceledException("Upload failed too many times.");
```

Comment:
Or just return an aggregate exception would be fine too

---
In reply to: [81231458](https://github.com/OneDrive/onedrive-sdk-csharp/pull/176#discussion_r81231458) [](ancestors = 81231458,81230799)